### PR TITLE
chore: upgrade node version to 18

### PR DESCRIPTION
**Pull request by yashasvibajpai**
fixed docker vulnerabilities and after fixing the below vulnerabilities remain only for both arm and amd architectures.

```
0C     0H     1M     0L  got 8.3.2
pkg:npm/got@8.3.2

Dockerfile (56:56)
RUN npm run clean:node

    ✗ MEDIUM CVE-2022-33987
      https://dso.docker.com/cve/CVE-2022-33987
      Affected range : <11.8.5                                       
      Fixed version  : 11.8.5                                        
      CVSS Score     : 5.3                                           
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:L/A:N  
    

   0C     0H     0M     0L     1?  cacheable-request 2.1.4
pkg:npm/cacheable-request@2.1.4

Dockerfile (56:56)
RUN npm run clean:node

    ✗ UNSPECIFIED GMS-2023-285
      https://dso.docker.com/cve/GMS-2023-285
      Affected range : <10.2.7  
      Fixed version  : 10.2.7   
    


2 vulnerabilities found in 2 packages
  UNSPECIFIED  1  
  LOW          0  
  MEDIUM       1  
  HIGH         0  
  CRITICAL     0

```

which previously  was :

```
Analyzing image rudderstack/rudder-transformer:1.27.1
    ✓ Pulled
    ✓ Image stored for indexing
    ✓ Indexed 1117 packages
    ✗ Detected 11 vulnerable packages with a total of 14 vulnerabilities

  1C    0H    0M    0L  execa 0.7.0
pkg:npm/execa@0.7.0

    ✗ CRITICAL GMS-2020-2
      https://dso.docker.com/cve/GMS-2020-2
      Affected range : <2.0.0
      Fixed version  : 2.0.0
      CVSS Score     : 9.8
      CVSS Vector    : CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H


  0C    1H    2M    2L  node 14.21.1
pkg:github/nodejs/node@14.21.1

    ✗ MEDIUM CVE-2023-23919 [Cryptographic Issues]
      https://dso.docker.com/cve/CVE-2023-23919
      Affected range : >=14.0.0
                     : <14.23.1
      Fixed version  : 14.23.1
      CVSS Score     : 7.5
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H

    ✗ HIGH CVE-2023-23918 [Incorrect Authorization]
      https://dso.docker.com/cve/CVE-2023-23918
      Affected range : >=14.0.0
                     : <14.23.1
      Fixed version  : 14.23.1
      CVSS Score     : 7.5
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:N/A:N

    ✗ LOW CVE-2023-24807 [Inefficient Regular Expression Complexity]
      https://dso.docker.com/cve/CVE-2023-24807
      Affected range : >=14.0.0
                     : <14.23.1
      Fixed version  : 14.23.1
      CVSS Score     : 7.5
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H

    ✗ MEDIUM CVE-2023-23936 [Improper Neutralization of Special Elements in Output Used by a Downstream Component ('Injection')]
      https://dso.docker.com/cve/CVE-2023-23936
      Affected range : >=14.0.0
                     : <14.23.1
      Fixed version  : 14.23.1
      CVSS Score     : 5.4
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:U/C:L/I:L/A:N

    ✗ LOW CVE-2023-23920 [Untrusted Search Path]
      https://dso.docker.com/cve/CVE-2023-23920
      Affected range : >=14.0.0
                     : <14.23.1
      Fixed version  : 14.23.1
      CVSS Score     : 4.2
      CVSS Vector    : CVSS:3.1/AV:L/AC:L/PR:H/UI:R/S:U/C:N/I:H/A:N


  0C    1H    1M    0L  decode-uri-component 0.2.0
pkg:npm/decode-uri-component@0.2.0

    ✗ HIGH CVE-2022-38900 [Improper Input Validation]
      https://dso.docker.com/cve/CVE-2022-38900
      Affected range : <0.2.1
      Fixed version  : 0.2.1
      CVSS Score     : 7.5
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H

    ✗ MEDIUM CVE-2022-38778 [Improper Input Validation]
      https://dso.docker.com/cve/CVE-2022-38778
      Affected range : <0.2.1
      Fixed version  : 0.2.1
      CVSS Score     : 6.5
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:L/UI:N/S:U/C:N/I:N/A:H


  0C    1H    1M    0L  node 14.21.1
pkg:generic/node@14.21.1

    ✗ HIGH CVE-2023-23918 [Incorrect Authorization]
      https://dso.docker.com/cve/CVE-2023-23918
      Affected range : >=14.0.0
                     : <14.21.3
      Fixed version  : 14.21.3
      CVSS Score     : 7.5
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:N/A:N

    ✗ MEDIUM CVE-2023-23920 [Untrusted Search Path]
      https://dso.docker.com/cve/CVE-2023-23920
      Affected range : >=14.0.0
                     : <14.21.3
      Fixed version  : 14.21.3
      CVSS Score     : 4.2
      CVSS Vector    : CVSS:3.1/AV:L/AC:L/PR:H/UI:R/S:U/C:N/I:H/A:N


  0C    1H    0M    0L  ansi-regex 4.1.0
pkg:npm/ansi-regex@4.1.0

    ✗ HIGH CVE-2021-3807 [Incorrect Comparison]
      https://dso.docker.com/cve/CVE-2021-3807
      Affected range : >=4.0.0
                     : <4.1.1
      Fixed version  : 4.1.1
      CVSS Score     : 7.5
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H


  0C    1H    0M    0L  ansi-regex 3.0.0
pkg:npm/ansi-regex@3.0.0

    ✗ HIGH CVE-2021-3807 [Incorrect Comparison]
      https://dso.docker.com/cve/CVE-2021-3807
      Affected range : >=3.0.0
                     : <3.0.1
      Fixed version  : 3.0.1
      CVSS Score     : 7.5
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H


  0C    1H    0M    0L  http-cache-semantics 3.8.1
pkg:npm/http-cache-semantics@3.8.1

    ✗ HIGH CVE-2022-25881 [Inefficient Regular Expression Complexity]
      https://dso.docker.com/cve/CVE-2022-25881
      Affected range : <4.1.1
      Fixed version  : 4.1.1
      CVSS Score     : 7.5
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H


  0C    1H    0M    0L  minimatch 3.0.4
pkg:npm/minimatch@3.0.4

    ✗ HIGH CVE-2022-3517 [Uncontrolled Resource Consumption]
      https://dso.docker.com/cve/CVE-2022-3517
      Affected range : <3.0.5
      Fixed version  : 3.0.5
      CVSS Score     : 7.5
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H


  0C    1H    0M    0L  qs 6.5.2
pkg:npm/qs@6.5.2

    ✗ HIGH CVE-2022-24999 [Improperly Controlled Modification of Object Prototype Attributes ('Prototype Pollution')]
      https://dso.docker.com/cve/CVE-2022-24999
      Affected range : >=6.5.0
                     : <6.5.3
      Fixed version  : 6.5.3
      CVSS Score     : 7.5
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H


  0C    0H    1M    0L  request 2.88.0
pkg:npm/request@2.88.0

    ✗ MEDIUM CVE-2023-28155 [Server-Side Request Forgery (SSRF)]
      https://dso.docker.com/cve/CVE-2023-28155
      Affected range : <=2.88.2
      Fixed version  : not fixed
      CVSS Score     : 6.1
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:C/C:L/I:L/A:N


  0C    0H    1M    0L  got 6.7.1
pkg:npm/got@6.7.1

    ✗ MEDIUM CVE-2022-33987
      https://dso.docker.com/cve/CVE-2022-33987
      Affected range : <11.8.5
      Fixed version  : 11.8.5
      CVSS Score     : 5.3
      CVSS Vector    : CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:L/A:N



The image contains 11 packages with one or more vulnerability for a total of 17 vulnerabilities
  LOW      | 2
  MEDIUM   | 6
  HIGH     | 8
  CRITICAL | 1
```
## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,15 +1,11 @@
...
 RUN apk update
 RUN apk upgrade
 
-# installing specific python version based on your previous configuration
-RUN apk add --no-cache tini python2
-# installing specific make version based on your previous configuration
-RUN apk add make=4.2.1-r2 --repository=http://dl-cdn.alpinelinux.org/alpine/v3.11/main
-# installing specific gcc version based on your previous configuration
-RUN apk add g++=9.3.0-r0 --repository=http://dl-cdn.alpinelinux.org/alpine/v3.11/main
+RUN apk add --no-cache tini make g++ python3
```
Comment (utsabc) [inline Dockerfile:8]: Why are we not specifying make and g++ version like before

File: Dockerfile [hunk 2]
```diff
@@ -1,15 +1,11 @@
 # syntax=docker/dockerfile:1.4
-FROM node:14.21.1-alpine3.15 AS base
+FROM node:18.15.0-alpine3.17 AS base
+ENV HUSKY 0
```
Comment (utsabc) [inline Dockerfile:3]: Whats the usage of these

File: Dockerfile-ut-func [hunk 3]
```diff
@@ -4,7 +4,7 @@ FROM ubuntu:20.04 AS base
 RUN apt-get update \
   && apt-get install -y curl make g++ \
   && apt-get install -y cpio \
-  && curl -sL https://deb.nodesource.com/setup_14.x | bash \
+  && curl -sL https://deb.nodesource.com/setup_18.x | bash \
```
Comment (utsabc) [inline Dockerfile-ut-func:7]: Lets not touch Dockerfile-ut-func (this for user transform we should have discussion with Services team before doing this)

File: package.json [hunk 4]
```diff
@@ -69,7 +69,7 @@
     "https-proxy-agent": "^5.0.1",
     "is": "^3.3.0",
     "is-ip": "^3.1.0",
-    "isolated-vm": "4.4.2",
+    "isolated-vm": "4.5.0",
```
Comment (utsabc) [inline package.json:72]: Also why are we updating this? Did we test this?

**Review by saikumarrs — DISMISSED** [review]

File: Dockerfile (hunk 1 above)
Comment (saikumarrs) [reply Dockerfile:8]: We don't need to be specific with those package versions anymore.

File: Dockerfile (hunk 2 above)
Comment (saikumarrs) [reply Dockerfile:3]: Husky is a pre-commit and setup hooks module. We're trying to disable it with this environment variable.

File: Dockerfile-ut-func (hunk 3 above)
Comment (saikumarrs) [reply Dockerfile-ut-func:7]: We've taken them into confidence for this project. They'll have to complete their side of testing before green-lighting the release.

File: package.json (hunk 4 above)
Comment (saikumarrs) [reply package.json:72]: The older version of the package is not compatible with Node v18. This version fixes that issue.

File: Dockerfile-ut-func (hunk 3 above)
Comment (saikumarrs) [reply Dockerfile-ut-func:7]: CC @Jayachand @krishna2020

File: Dockerfile (hunk 1 above)
Comment (utsabc) [reply Dockerfile:8]: Ok majorly these were need as part of isolate vm I see we are updating python 3 (should be fine if we are testing these)

File: Dockerfile-ut-func (hunk 3 above)
Comment (utsabc) [reply Dockerfile-ut-func:7]: Ok make sense

File: package.json (hunk 4 above)
Comment (utsabc) [reply package.json:72]: Got it (lets make sure we specifically test scenarios for User transform in such case)

**Comment by saikumarrs** [comment]
Snyk scan report.

```
> snyk container test rudderstack/develop-rudder-transformer:chore.upgrade.node.to.18 --file=./Dockerfile

Testing rudderstack/develop-rudder-transformer:chore.upgrade.node.to.18...

Organization:      saikumarrs
Package manager:   apk
Target file:       ./Dockerfile
Project name:      docker-image|rudderstack/develop-rudder-transformer
Docker image:      rudderstack/develop-rudder-transformer:chore.upgrade.node.to.18
Platform:          linux/arm64
Base image:        node:18.15.0-alpine3.17
Licenses:          enabled

✔ Tested 43 dependencies for known issues, no vulnerable paths found.

According to our scan, you are currently using the most secure version of the selected base image

-------------------------------------------------------

Testing rudderstack/develop-rudder-transformer:chore.upgrade.node.to.18...

Tested 56 dependencies for known issues, found 1 issue.


Issues with no direct upgrade or patch:
  ✗ Remote Code Execution (RCE) [Critical Severity][https://security.snyk.io/vuln/SNYK-JS-ISOLATEDVM-3037320] in isolated-vm@4.5.0
    introduced by isolated-vm@4.5.0
  No upgrade or patch available



Organization:      saikumarrs
Package manager:   npm
Target file:       /home/node/app/package.json
Project name:      rudder-transformer
Docker image:      rudderstack/develop-rudder-transformer:chore.upgrade.node.to.18
Licenses:          enabled

Pro tip: use `--exclude-base-image-vulns` to exclude from display Docker base image vulnerabilities.

Snyk found some vulnerabilities in your image applications (Snyk searches for these vulnerabilities by default). See https://snyk.co/app-vulns for more information.

To remove these messages in the future, please run `snyk config set disableSuggestions=true`


Tested 2 projects, 1 contained vulnerable paths.
```

We can suppress the above `isolated-vm` vulnerability as we don't seem to use that feature.
CC @Jayachand @sivashanmukh @krishna2020

**Review by ItsSudip — APPROVED** [review]

**Review by shrouti1507 — APPROVED** [review]

**Review by sanpj2292 — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**